File: cpp/questions/sortColors.cpp
#include <bits/stdc++.h>

using namespace std;

class Solution {
public:
    void sortColors(vector<int>& nums) {
        int ns = nums.size();
        int l = 0;
        int r = ns - 1;
        int p = l;
        while (p <= r) {
            if (nums[p] == 0) {
                std::swap(nums[p], nums[l]);
                while (l<ns && nums[l]==0) {
                    ++l;
                }
                p = std::max(l, p);
            } else if (nums[p] == 2) {
                std::swap(nums[p], nums[r]);
                while (r>=0 && nums[r]==2) {
                    --r;
                }
            }
            p += p<ns && nums[p]==1 ? 1 : 0;
        }
    }
};

File: python/questions/sortColors.py
from typing import List

class Solution:
    def sortColors(self, nums: List[int]) -> None:
        ln = len(nums)
        l = 0
        r = ln - 1
        p = l
        while p <= r:
            if nums[p] == 0:
                nums[l], nums[p] = nums[p], nums[l]
                while l<ln and nums[l]==0:
                    l += 1
                p = max(l, p)
            elif nums[p] == 2:
                nums[r], nums[p] = nums[p], nums[r]
                while r>=0 and nums[r]==2:
                    r -= 1
            p += 1 if p<ln and nums[p]==1 else 0
